### DISCUSSION TITLE
Using std::optional for optional matrix arguments

### DISCUSSION THREAD

**Discussion by thijssteel** [tlapack/tlapack#218] [Ideas]
When implementing the QR factorization, I looked at some ways to use an optional argument for `TT` the matrix containing the triangular factors of the WY representations. But those never really worked nicely.

I just stumbled onto another way: `std::optional`. It does not work perfectly (in particular, it struggles with template deduction of types that were not provided).

I made a small example emulating the svd implementation I was working on. Normally, you would have an interface like this:

```c++
template<class T>
int gesvd(bool want_u, bool want_vt, T& A, T& s, T& U, T& Vt){
    return 0;
}
```
where `U` and `Vt` would not be referenced if `want_u == false` and/or `want_vt == false`. However, it still requires that you pass an empty matrix as argument and I don't really like that code.

```c++
#include <optional>
 
// This function contains the actual implementation
// Note the lack of a parameter like want_u or want_vt, if U or Vt are provided
// we can assume they are desired.
template<class T>
int gesvd_impl(T& A, T& s, const std::optional<T>& U, const std::optional<T>& Vt){
    return 0;
}

// 
// std::optional does not work very well with templates so we make some wrappers
// 

template<class T>
int gesvd(T& A, T& s, T& U, T& Vt){
    return gesvd_impl<T>(A,s,std::make_optional(U),std::make_optional(Vt));
}

template<class T>
int gesvd(T& A, T& s, T& U){
    // Note the use of explicit templates, otherwise the deduction fails.
    return gesvd_impl<T>(A,s,std::make_optional(U),std::nullopt);
}

template<class T>
int gesvd(T& A, T& s, std::nullopt_t U, T& Vt){
    return gesvd_impl<T>(A,s,std::nullopt,std::make_optional(Vt));
}

template<class T>
int gesvd(T& A, T& s){
    return gesvd_impl<T>(A,s,std::nullopt,std::nullopt);
}

int main()
{
    float A, s, U, Vt;
    // Look how nice the interface is for the user
    gesvd(A, s, U, Vt);
    gesvd(A, s, std::nullopt, Vt);
    gesvd(A, s, U);
    gesvd(A, s);
}

```

What do you think?

**Comment by thijssteel** [comment]
I just noticed a flaw in this design. If you want to propagate those optional arguments, you would run into problems.

**Comment by weslleyspereira** [comment]
I also don't like passing arguments and not using them inside the routine. Thanks for raising this discussion!

What about having something like the following?

```c++
template<class T, bool want_u, bool want_vt>
int gesvd_impl(T& A, T& s, T& U, T& Vt){
    return 0;
}

template<class T>
int gesvd(T& A, T& s, T& U, T& Vt){
    return gesvd_impl<T,true,true>(A,s,U,Vt);
}

template<class T>
int gesvd_U(T& A, T& s, T& U){
    // create empty Vt
    return gesvd_impl<T,true,false>(A,s,U,Vt);
}

template<class T>
int gesvd_Vt(T& A, T& s, T& Vt){
    // create empty U
    return gesvd_impl<T,true,false>(A,s,U,Vt);
}

template<class T>
int gesvd(T& A, T& s){
    // create empty U
    // create empty Vt
    return gesvd_impl<T,false,false>(A,s,U,Vt);
}
```

**Reply by langou** [reply]
I am slightly worried about combinatorial explosion here. If you have three optional arguments, and they all are independent, either on or off. Then that would 2^3 = 8 interfaces. Me too, I am not a big fan of passing arguments that are not used. The design of @weslleyspereira looks good to me. We can try it like that.

**Comment by thijssteel** [comment]
I also like weslleys design. Creating an empty matrix is a bit problematic though. If created using the create_matrix functor, some types like eigen complain about null pointers. Maybe we need a special dummy matrix class.
